musig/src/signer.rs

```diff
@@ -0,0 +1,258 @@
+            }
+        }        
+
+        // R = \sum{1<=i<=n}{R_i}
+        let mut acc = Point::zero();
+        for commitment in commitments{
+            acc = self.jubjub_wrapper.add(&acc, &commitment);
+        }
+        self.aggregated_commitment = Some(acc.clone());
+        self.nonce_commitments = commitments.to_vec();
+
+        Ok(acc)
+    }
+
+    pub fn sign(
+        &mut self,
+        private_key: &PrivateKey<E>,
+        message: &[u8],
+        rescue_params: &<E as RescueEngine>::Params,
+    ) -> Result<E::Fs, MusigError>{      
+        // check that whether previous step passed or not        
+        if self.aggregated_commitment.is_none(){
+            return Err(MusigError::NonceCommitmentsNotReceived)
+        }
+
+        let aggregated_commitment = self.aggregated_commitment
+            .clone()
+            .unwrap();
+        // since aggregated commitment has already generated 
+        // we can safely unwrap nonce
+        let r = self.nonce.unwrap(); 
+                    
+        let aggregated_pubkey = self.aggregated_pubkey.clone();
+
+        let a_i = self.a_values[self.position].clone();
+
+        // c = H_sig(X', R, m)
+        // this computes fiat-shamir challenge
+        let c = Hasher::hash_signature_data::<E>(
+            &aggregated_pubkey, 
+            &aggregated_commitment, 
+            message,
+            rescue_params
+        );
+
+        self.challenge = Some(c);
+        // s = r + c * a_i * x_i  
+        let mut s = c.clone();
+        s.mul_assign(&a_i);
+        s.mul_assign(&private_key.0);
+        s.add_assign(&r);
+
+        Ok(s)
+    }
+
+    pub fn receive_signatures(
+        &self,
+        signature_shares: &[E::Fs],
+    ) -> Result<Signature<E>, MusigError> {        
+        // check that whether previous step passed or not
+        if self.challenge.is_none(){
+            return Err(MusigError::ChallengeNotGenerated)
+        }
+
+        if signature_shares.len() != self.pubkeys.len(){
+            return Err(MusigError::SignatureShareAndParticipantsNotMatch);
+        }
+
+        let mut aggregated_signature = self.signature.clone();     
+        // s = \sum{1<=i<=n}{s_i}
+        for (position, signature) in signature_shares.iter().enumerate(){
+            // verify each signature share
+            // s*G = R_i + (c * a_i) * X_i   
+            self.verify_share(signature, position)?;
+            aggregated_signature.add_assign(&signature);
+        }
+
+        let aggregated_commitment = self.aggregated_commitment
+        .clone()
+        .unwrap();  
+
+        // σ = (R, s)
+        Ok(Signature{
+            r: aggregated_commitment, 
+            s: aggregated_signature,
+        })
+    }    
+
+    fn verify_share(
+        &self,
+        signature_share: &E::Fs,
+        position: usize,
+    ) -> Result<(), MusigError>{
+        let challenge = self.challenge
+            .clone()
```

Comment:
As you only need `challenge` as a reference, I believe you can avoid cloning it.